e2e_test.go

```diff
@@ -0,0 +1,313 @@
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	metricTimestamp = time.Date(2020, 2, 11, 20, 26, 12, 321, time.UTC)
+)
+
+// newTs advances the metricTimestamp and returns a new timestamp.
+func newTs() pcommon.Timestamp {
+	metricTimestamp = metricTimestamp.Add(time.Minute)
+	return pcommon.NewTimestampFromTime(metricTimestamp)
+}
+
+// configProviderSettings is a convenience function to create ConfigProviderSettings that use the
+// local config.yaml.
+func configProviderSettings() otelcol.ConfigProviderSettings {
+	return otelcol.ConfigProviderSettings{
+		ResolverSettings: confmap.ResolverSettings{
+			URIs: []string{filepath.Join("config", "config.yaml")},
+			ProviderFactories: []confmap.ProviderFactory{
+				fileprovider.NewFactory(),
+			},
+		},
+	}
+}
+
+// subscribeRequestForTarget is a convenience function to create a target-specific SubscribeRequest.
+func subscribeRequestForTarget(t *testing.T, target string) *gpb.SubscribeRequest {
+	return &gpb.SubscribeRequest{
+		Request: &gpb.SubscribeRequest_Subscribe{
+			Subscribe: &gpb.SubscriptionList{
+				Prefix: &gpb.Path{
+					Target: target,
+					Elem: []*gpb.PathElem{
+						{
+							Name: target,
+						},
+					},
+				},
+				Mode: gpb.SubscriptionList_STREAM,
+				Subscription: []*gpb.Subscription{
+					&gpb.Subscription{
+						Mode:              gpb.SubscriptionMode_ON_CHANGE,
+						Path:              &gpb.Path{},
+						SuppressRedundant: false,
+					},
+				},
+			},
+		},
+	}
+}
+
+// startCollectorPipeline starts and returns the collector pipeline (plus the WaitGroup it runs in).
+func startCollectorPipeline(t *testing.T, ctx context.Context) (*sync.WaitGroup, *otelcol.Collector) {
+	t.Helper()
+	t.Log("Starting collector pipeline")
+
+	// Create collector.
+	set := otelcol.CollectorSettings{
+		BuildInfo:              component.NewDefaultBuildInfo(),
+		Factories:              collector.Components,
+		ConfigProviderSettings: configProviderSettings(),
+	}
+	col, err := otelcol.NewCollector(set)
+	require.NoError(t, err)
+
+	// Start collector.
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		require.NoError(t, col.Run(ctx))
+	}()
+	return wg, col
+}
+
+// stopCollectorPipeline stops the collector pipeline and waits for it to finish.
+func stopCollectorPipeline(t *testing.T, wg *sync.WaitGroup, col *otelcol.Collector) {
+	t.Helper()
+	t.Log("Stopping collector pipeline")
+	col.Shutdown()
+	wg.Wait()
+	t.Log("Stopped collector pipeline")
+	assert.Equal(t, otelcol.StateClosed, col.GetState())
+}
+
+func TestE2E(t *testing.T) {
+	// Number of metric waves to export. Based on the output of GenerateMetrics(...),
+	// each wave should produce 20 notifications.
+	exportWaves := 10
+
+	gOpts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+	ctx := context.Background()
+
+	// Early declaration of wait group to ensure that defer order works.
+	sinkWg := &sync.WaitGroup{}
+	defer sinkWg.Wait()
```

Comment:
Does this need to be deferred or could it happen before the function exits?

Reply:
The sink wait for receiving an error (EOF or rpc code canceled) via the stream from the source (the gnmi server in the collector). If this is executed without deferring (at that exact position), then the sink will be stuck in the stream.Recv() call as the collector isn't shut down yet but also does not send any data.